Improve Search Box Speed to make it more "buttery" + type anywhere to search.

wenqinYe opened the pull request:
The live emoji search is great, expect for one problem; it slowed down the system as it made a separate search call for every character typed. This would result in noticeable lag in typing (this is even a problem on my MBP 2012).
So what I did was put a delay on the search so that the search is performed when the user is done typing. This way, there are no useless (and expensive) search calls per character; instead, there is only one call at the end when the user finished typing. 

Of course, one thing that live search was great at doing was giving the feeling of instantaneous search results. I wanted to keep this illusion of "instant" search by making the results disappear to give the user the sense that their search is being instantly acknowledged. Without this, the user is stuck on the previous results page until they finish typing, which can give a feeling of sluggishness from the system.

---

As a side, you can now type anywhere to search; I removed the feature of having to type in a slash first. I don't think people want to type in an arbitrary command when their intent is to access the search bar and find some awesome emojis :smile:. They should be able to access the search bar right away.

Cheers
:beers:

Commit:
Added Typing Delay For illusion of smoother experience

Commit:
Faster Search Speed and Enter To Search Anywhere

Commit:
added comments

Commit:
Updated Delay Time and Comments

wenqinYe committed:
update delay time

brianchung808 commented:
You should make the changes in the .coffee files since the js you edited is just generated.

wenqinYe commented:
Ahh thank you I was not aware that we were supposed to edit the .coffee files.

Commit:
rever committ

Commit:
Revert "added comments"

This reverts commit 472ca93f7db213ff631cee9d6b17d245abb28be2.

Conflicts:
	javascripts/speedy.js

Commit:
Revert 47ac2c1..f526dff

This rolls back to commit 47ac2c12cfe73d313ce68e4277d8cd07bfca58ea.

Commit:
Improve Search in coffee script file

changed js last time, finally got it into coffee script after a couple
days of doing other things.

Commit:
added typing anywhere feature

wenqinYe commented:
Sorry for the delay but today I switched everything to coffee script.
Could muan please look at this commit; thank you!

Also I decided to keep the "live" search updating instead of everything going away when the user types something and then shows results when the user finishes typing.

Cheers
:beers:

brianchung808 commented on _assets/speedy.coffee:
```diff
@@ -30,8 +30,27 @@ setRelatedDOMVisibility = (keyword) ->
+  clearTimeout(timeout);
+
+  timeout = setTimeout(
+    (()-> search($('.speedy-filter').val()) ), delayTime)
```
style-wise it'd be nicer to omit the `()` in the `() ->` and leave it as `->`. (https://github.com/styleguide/javascript)

wenqinYe replied:
Ok thank you brian :)

brianchung808 commented:
The semicolons should be removed (:mask:) but I do like being able to just start typing to search rather than the vim-y `/` method :+1:

brianchung808 committed:
fix focusOnSearch

It will now actually focus on the search bar when you hit a key

brianchung808 commented:
@wenqinYe I downloaded your fork & the `focusOnSearch` doesn't actually focus on the search bar. I sent a pull req with a fix

wenqinYe committed:
Merge pull request #1 from brianchung808/patch-1

fix focusOnSearch

wenqinYe committed:
removed unnecessary comment

Commit:
Improve style + hash was not being set when searching

Commit:
Merge branch 'gh-pages' of https://github.com/wenqinYe/emoji into gh-pages

muan commented:
Hai, sorry for the delay in replying! 
@brianchung808 thanks so much for help moving this forward while I be lazy!

> Also I decided to keep the "live" search updating instead of everything going away when the user types something and then shows results when the user finishes typing.

Just pulled it down, looks like the search is not "live"? do you mean that you'd keep the delay but don't clear the previous results? It feels pretty slow to me:

![](http://cl.ly/image/3P1k1d0O0b1h/Screen%20Recording%202015-01-19%20at%2012.25%20PM.gif)

Compared to the current behavior:

![](http://cl.ly/image/3E411g2J0r1K/Screen%20Recording%202015-01-19%20at%2012.32%20PM.gif)

wenqinYe commented:
I originally planned on clearing the previous search results as soon as the user typed in a search request to give the illusion of speed, but I changed it because I thought you might not have wanted that.

muan commented:
Thanks! I like being able to type/search anywhere.

But regarding the performance issue, given that the lag doesn't exist on every machine(on my MBP it's pretty snappy), the solution here seems to be making search considerably slower for everyone, which isn't great. Have you tried cutting the wait time down to something closer to time between keystrokes?

wenqinYe commented:
I'm going to work on this again after exams. Been too busy studying right now, will comment back when I have a new commit for you.

muan commented:
Closing as stale.
